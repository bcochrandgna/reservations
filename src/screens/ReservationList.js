// @flow
import React, { PureComponent } from 'react';
import { SafeAreaView } from 'react-native';
import { Query } from 'react-apollo';
import { Icon } from 'react-native-elements';
import styles from '../styles/screens/ReservationList';
import GET_RESERVATIONS from '../data/queries/getReservations';
import InfiniteDataList from '../components/InfiniteDataList';
import ReservationCard from '../components/ReservationCard';
import ReservationListHeader from '../components/ReservationListHeader';
import ErrorIndicator from '../components/ErrorIndicator';
import GET_SORT_OPTIONS from '../data/queries/getSortOptions';
import i18n from '../helpers/i18n';
import { Reservation } from '../types/Reservation';

type Props = {};
type TabBarIconArguments = {
  tintColor: string,
};

class ReservationList extends PureComponent<Props> {
  static navigationOptions = {
    tabBarIcon: ({ tintColor }: TabBarIconArguments) => <Icon name="list" color={tintColor} />,
    title: i18n.t('reservations'),
  };

  renderItem = (item: Reservation) => <ReservationCard reservation={item} />;

  render() {
    return (
      <SafeAreaView style={styles.container}>
        <ReservationListHeader />
        <Query query={GET_SORT_OPTIONS}>
          {({ data: { sortOptions } }) => (
            <Query
              query={GET_RESERVATIONS}
              variables={{
                skip: 0,
                first: 10,
                orderBy: sortOptions,
              }}
              fetchPolicy="cache-and-network"
              notifyOnNetworkStatusChange
            >
              {({
                data, error, networkStatus, fetchMore, refetch,
              }) => {
                const reservations = data && data.reservations;

                if (error) {
                  return <ErrorIndicator error={error.message} />;
                }

                return (
                  <InfiniteDataList
                    data={reservations}
                    networkStatus={networkStatus}
                    fetchMore={() => {
                      fetchMore({
                        variables: {
                          skip: reservations.length,
                        },
                        updateQuery: (previousResult, { fetchMoreResult }) => {
                          if (!fetchMoreResult || fetchMoreResult.reservations.length === 0) {
                            return previousResult;
                          }

                          return {
                            reservations: previousResult.reservations.concat(
                              fetchMoreResult.reservations,
                            ),
                          };
                        },
                      });
                    }}
                    renderItem={this.renderItem}
                    refetch={refetch}
                  />
                );
              }}
            </Query>
          )}
        </Query>
      </SafeAreaView>
    );
  }
}

export default ReservationList;
