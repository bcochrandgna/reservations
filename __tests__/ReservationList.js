import 'react-native';
import React from 'react';
import { MockedProvider } from 'react-apollo/test-utils';
import renderer from 'react-test-renderer';
import wait from 'waait';
import ReservationList from '../src/screens/ReservationList';
import mocks from '../fixtures/mocks';

describe('ReservationList', () => {
  it('Renders properly', async () => {
    const wrapper = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ReservationList />
      </MockedProvider>,
    );

    await wait(100);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
