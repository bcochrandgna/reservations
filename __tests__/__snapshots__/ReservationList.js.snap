// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReservationList Renders properly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#007AFF",
          "fontSize": 18,
        }
      }
    >
      Reservations (0)
    </Text>
    <View
      style={Object {}}
    >
      <Modal
        animationType="slide"
        hardwareAccelerated={false}
        onDismiss={[Function]}
        onRequestClose={[Function]}
        supportedOrientations={
          Array [
            "portrait",
            "landscape",
          ]
        }
        transparent={true}
        visible={false}
      >
        <View
          accessible={false}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "rgba(0,0,0,0.7)",
                "flex": 1,
                "justifyContent": "center",
                "padding": "5%",
              },
              Object {},
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "rgba(255,255,255,0.8)",
                  "borderRadius": 5,
                  "flexShrink": 1,
                  "marginBottom": 8,
                  "padding": 8,
                },
                Object {},
              ]
            }
          >
            <RCTScrollView
              accessible={false}
              keyboardShouldPersistTaps="always"
            >
              <View>
                <View
                  style={
                    Object {
                      "paddingHorizontal": 10,
                    }
                  }
                >
                  <View
                    accessible={false}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderBottomColor": "#ccc",
                            "borderBottomWidth": 1,
                            "padding": 8,
                          },
                          Object {},
                          false,
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "rgba(0,118,255,0.9)",
                              "fontSize": 16,
                              "textAlign": "center",
                            },
                            Object {},
                            false,
                          ]
                        }
                      >
                        Name (ASC)
                      </Text>
                    </View>
                  </View>
                  <View
                    accessible={false}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderBottomColor": "#ccc",
                            "borderBottomWidth": 1,
                            "padding": 8,
                          },
                          Object {},
                          false,
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "rgba(0,118,255,0.9)",
                              "fontSize": 16,
                              "textAlign": "center",
                            },
                            Object {},
                            false,
                          ]
                        }
                      >
                        Name (DESC)
                      </Text>
                    </View>
                  </View>
                  <View
                    accessible={false}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderBottomColor": "#ccc",
                            "borderBottomWidth": 1,
                            "padding": 8,
                          },
                          Object {},
                          false,
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "rgba(0,118,255,0.9)",
                              "fontSize": 16,
                              "textAlign": "center",
                            },
                            Object {},
                            false,
                          ]
                        }
                      >
                        Hotel Name (ASC)
                      </Text>
                    </View>
                  </View>
                  <View
                    accessible={false}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderBottomColor": "#ccc",
                            "borderBottomWidth": 1,
                            "padding": 8,
                          },
                          Object {},
                          false,
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "rgba(0,118,255,0.9)",
                              "fontSize": 16,
                              "textAlign": "center",
                            },
                            Object {},
                            false,
                          ]
                        }
                      >
                        Hotel Name (DESC)
                      </Text>
                    </View>
                  </View>
                  <View
                    accessible={false}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderBottomColor": "#ccc",
                            "borderBottomWidth": 1,
                            "padding": 8,
                          },
                          Object {},
                          false,
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "rgba(0,118,255,0.9)",
                              "fontSize": 16,
                              "textAlign": "center",
                            },
                            Object {},
                            false,
                          ]
                        }
                      >
                        Arrival Date (ASC)
                      </Text>
                    </View>
                  </View>
                  <View
                    accessible={false}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderBottomColor": "#ccc",
                            "borderBottomWidth": 1,
                            "padding": 8,
                          },
                          Object {},
                          false,
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "rgba(0,118,255,0.9)",
                              "fontSize": 16,
                              "textAlign": "center",
                            },
                            Object {},
                            false,
                          ]
                        }
                      >
                        Arrival Date (DESC)
                      </Text>
                    </View>
                  </View>
                  <View
                    accessible={false}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderBottomColor": "#ccc",
                            "borderBottomWidth": 1,
                            "padding": 8,
                          },
                          Object {},
                          false,
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "rgba(0,118,255,0.9)",
                              "fontSize": 16,
                              "textAlign": "center",
                            },
                            Object {},
                            false,
                          ]
                        }
                      >
                        Departure Date (ASC)
                      </Text>
                    </View>
                  </View>
                  <View
                    accessible={false}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderBottomColor": "#ccc",
                            "borderBottomWidth": 1,
                            "padding": 8,
                          },
                          Object {},
                          Object {
                            "borderBottomWidth": 0,
                          },
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "rgba(0,118,255,0.9)",
                              "fontSize": 16,
                              "textAlign": "center",
                            },
                            Object {},
                            false,
                          ]
                        }
                      >
                        Departure Date (DESC)
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </RCTScrollView>
          </View>
          <View
            style={
              Array [
                Object {
                  "alignSelf": "stretch",
                },
                Object {},
              ]
            }
          >
            <View
              accessible={false}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "backgroundColor": "rgba(255,255,255,0.8)",
                      "borderRadius": 5,
                      "padding": 8,
                    },
                    Object {},
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#333",
                        "fontSize": 16,
                        "textAlign": "center",
                      },
                      Object {},
                    ]
                  }
                >
                  cancel
                </Text>
              </View>
            </View>
          </View>
        </View>
      </Modal>
      <View
        accessible={false}
        hitSlop={
          Object {
            "bottom": 0,
            "left": 0,
            "right": 0,
            "top": 0,
          }
        }
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          pointerEvents="none"
          style={Object {}}
        >
          <View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
              theme={
                Object {
                  "colors": Object {
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "primary": "#2196f3",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "primary": "#007aff",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#8F0CE8",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                }
              }
              underlayColor="white"
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "black",
                      "fontSize": 24,
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                    Object {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
                testID="iconIcon"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "arrivalDate": "03/01/2019",
          "departureDate": "03/03/2019",
          "hotelName": "Test Hotel 1",
          "id": "1",
          "name": "Test Name 1",
        },
        Object {
          "arrivalDate": "03/01/2019",
          "departureDate": "03/03/2019",
          "hotelName": "Test Hotel 2",
          "id": "2",
          "name": "Test Name 2",
        },
        Object {
          "arrivalDate": "03/01/2019",
          "departureDate": "03/03/2019",
          "hotelName": "Test Hotel 3",
          "id": "3",
          "name": "Test Name 3",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#fafafa",
              "borderColor": "#dddddd",
              "borderWidth": 1,
              "marginBottom": 5,
              "marginHorizontal": 5,
            }
          }
        >
          <View
            style={
              Object {
                "padding": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "red",
                  "fontSize": 16,
                }
              }
            >
              Test Name 1
            </Text>
            <Text>
              Test Hotel 1
            </Text>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#f0f0f0",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginTop": 5,
                "padding": 5,
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontWeight": "700",
                    }
                  }
                >
                  Arrival:
                </Text>
                <Text
                  style={
                    Object {
                      "marginLeft": 5,
                    }
                  }
                >
                  Friday, March 1st 2019
                </Text>
              </View>
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontWeight": "700",
                    }
                  }
                >
                  Departure:
                </Text>
                <Text
                  style={
                    Object {
                      "marginLeft": 5,
                    }
                  }
                >
                  Sunday, March 3rd 2019
                </Text>
              </View>
            </View>
            <View />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#fafafa",
              "borderColor": "#dddddd",
              "borderWidth": 1,
              "marginBottom": 5,
              "marginHorizontal": 5,
            }
          }
        >
          <View
            style={
              Object {
                "padding": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "red",
                  "fontSize": 16,
                }
              }
            >
              Test Name 2
            </Text>
            <Text>
              Test Hotel 2
            </Text>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#f0f0f0",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginTop": 5,
                "padding": 5,
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontWeight": "700",
                    }
                  }
                >
                  Arrival:
                </Text>
                <Text
                  style={
                    Object {
                      "marginLeft": 5,
                    }
                  }
                >
                  Friday, March 1st 2019
                </Text>
              </View>
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontWeight": "700",
                    }
                  }
                >
                  Departure:
                </Text>
                <Text
                  style={
                    Object {
                      "marginLeft": 5,
                    }
                  }
                >
                  Sunday, March 3rd 2019
                </Text>
              </View>
            </View>
            <View />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#fafafa",
              "borderColor": "#dddddd",
              "borderWidth": 1,
              "marginBottom": 5,
              "marginHorizontal": 5,
            }
          }
        >
          <View
            style={
              Object {
                "padding": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "red",
                  "fontSize": 16,
                }
              }
            >
              Test Name 3
            </Text>
            <Text>
              Test Hotel 3
            </Text>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#f0f0f0",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginTop": 5,
                "padding": 5,
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontWeight": "700",
                    }
                  }
                >
                  Arrival:
                </Text>
                <Text
                  style={
                    Object {
                      "marginLeft": 5,
                    }
                  }
                >
                  Friday, March 1st 2019
                </Text>
              </View>
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontWeight": "700",
                    }
                  }
                >
                  Departure:
                </Text>
                <Text
                  style={
                    Object {
                      "marginLeft": 5,
                    }
                  }
                >
                  Sunday, March 3rd 2019
                </Text>
              </View>
            </View>
            <View />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
